File: medium/54.ts
function spiralOrder(matrix: number[][]): number[] {
    const res: number[] = []
    dfs(matrix)
    return res

    function dfs(submat: number[][]){
        const m = submat.length
        const n = submat[0].length

        if(m == 1){
            submat[0].forEach(v => res.push(v))
        }
        else if(n == 1) {
            for(let i = 0; i < m; ++i)
                res.push(submat[i][0])
        }
        else{
            for(let i = 0; i < n; ++i)
                res.push(submat[0][i])
            for(let i = 1; i < m; ++i)
                res.push(submat[i][n-1])
            for(let i = n - 2; i > 0; --i)
                res.push(submat[m - 1][i])
            for(let i = m - 1; i > 0; --i)
                res.push(submat[i][0])
                
            if(m > 2 && n > 2){
                const next: number[][] = []
                for(let i = 1; i < m - 1; ++i)
                    next.push(submat[i].slice(1, n - 1))
                dfs(next)
            }
        }
    }
};

File: medium/54.cpp
/**
    # 54. Spiral Matrix
*/
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int t = 0, b = matrix.size() - 1, l = 0, r = matrix[0].size() - 1;
        int dir[8] = {0, 1, 1, 0, 0, -1, -1, 0};
        int k = 0;
        int x = 0, y = 0;
        vector<int> res;
        
        while(l <= r && t <= b){
            res.push_back(matrix[x][y]);
            
            switch(k){
                case 0: 
                    if(y >= r) {
                        k = 1;
                        ++t;
                    }
                    break;
                case 1: 
                    if(x >= b){
                        k = 2;
                        --r;
                    }
                    break;
                case 2:
                    if(y <= l){
                        k = 3;
                        --b;
                    }
                    break;
                case 3:
                    if(x <= t){
                        k = 0;
                        ++l;
                    }
                    break;
            }
            x += dir[k * 2];
            y += dir[k * 2 + 1];
        }
        
        return res;
    }
};
